File: metodosOrdenacao/CPP/problema.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <ctime>
#include "problema.hpp"

using namespace std;

void ler_input(problema* prob, string filename){
    ifstream file(filename);
    if (!file.is_open()) {
        cout << "\033[31m Erro ao abrir arquivo\n";
        cout <<" Arquivo não encontrado ou inacessível: "<< filename <<"\n \033[0m";
        exit(1);
    }

    int tamanho = 0;
    file >> tamanho;
    vector<int> lista(tamanho);

    int valor;
    for(int i = 0; i < tamanho; file >> lista[i++]);

    prob->lista = lista;

    file.close();
}

string obter_nome_input(int input, string nome_input){
    if (input == ESCOLHER_EM_EXECUCAO) {
        do{
            cout << " Escolha o Input a ser usado:\n";
            cout << "1. Input 1\n2. Input 2\n3. Input 3\n4. Input 4\n5. Input 5\n6. Input 6\nResposta: ";
            
            cin >> input;
            
            if (input < 1 || input > 6) cout << "\n \033[31m Opção inválida. Por favor, escolha um número entre 1 e 6. \033[0m \n\n";
        } while (input < 1 || input > 6);

        switch (input) {
            case 1: nome_input = INPUT1; break;
            case 2: nome_input = INPUT2; break;
            case 3: nome_input = INPUT3; break;
            case 4: nome_input = INPUT4; break;
            case 5: nome_input = INPUT5; break;
            case 6: nome_input = INPUT6; break;
        }
    }

    return nome_input;
}

void obter_algoritmo_usado(int algoritmo, int* algoritmo_usado) {
    if (algoritmo == ESCOLHER_EM_EXECUCAO) {
        do {
            cout << "\n Escolha o algoritmo:\n1. Radix\n2. Counting\n3. intro Sort\n4. Todos\nResposta: ";
            
            cin >> algoritmo;
            if (algoritmo < 1 || algoritmo > 4)
                cout << "\n \033[31m Opção inválida.\033[0m\n\n";
        } while (algoritmo < 1 || algoritmo > 4);
    }
    *algoritmo_usado = algoritmo; // ✅ salva o valor
}

void obter_quantidade_execucoes(int quantidade_execucoes, int* quantidade_execucoes_obtida) {
    if (quantidade_execucoes == ESCOLHER_EM_EXECUCAO) {
        do {
            cout << "\n Digite a quantidade de execuções (1 ou mais): ";
            
            cin >> quantidade_execucoes;
            if (quantidade_execucoes < 1)
                cout << "\n \033[31m Quantidade inválida.\033[0m\n\n";
        } while (quantidade_execucoes < 1);
    }
    *quantidade_execucoes_obtida = quantidade_execucoes; // ✅ salva o valor
}

problema* criar_problema(int input, int algoritmo, int quantidade_execucoes){
    problema* prob = new problema;
    prob->algoritmo_usado = algoritmo;
    prob->quantidade_execucoes = quantidade_execucoes;

    prob->nome_input = obter_nome_input(input, prob->nome_input);
    obter_algoritmo_usado(algoritmo, &prob->algoritmo_usado);
    obter_quantidade_execucoes(quantidade_execucoes, &prob->quantidade_execucoes);
    ler_input(prob, prob->nome_input);

    return prob;
}

void criar_output(problema* prob, vector<double> lista_tempos){
    int nome_algoritmo = prob->algoritmo_usado;
    int quantidade_execucoes = prob->quantidade_execucoes;
    string nome_input = prob->nome_input;

    string nome_algoritmo_str;

    switch (nome_algoritmo) {
        case RADIX_SORT: nome_algoritmo_str = "Radix Sort"; break;
        case COUNTING_SORT: nome_algoritmo_str = "Counting Sort"; break;
        case INTRO_SORT: nome_algoritmo_str = "introsort Sort"; break;
    }

    string nome_arquivo = "outputs/output"  + nome_algoritmo_str + to_string(time(NULL)) + ".dat";

    ofstream file(nome_arquivo);

    if (!file.is_open()) {
        cout << "Erro ao criar o arquivo de saída.\n";
        return;
    }

    file << "Linguagem: C++" << endl;
    file << "Algoritmo: " << nome_algoritmo_str << endl;
    file << "Input: " << nome_input << endl;

    file << "Tempos de execução (em segundos):" << endl;;
    for (int i = 0; i < quantidade_execucoes; file << "Execução " <<  i << ": "<< lista_tempos[i++] << endl);
    
    file.close();
}

void exibir_lista(vector<int> lista){
    for (int valor : lista){
        cout << valor << " ";
    }
}

File: metodosOrdenacao/RUST/problema.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Write};
use std::process;
use std::time::{SystemTime, UNIX_EPOCH};

pub const ESCOLHER_EM_EXECUCAO: i32 = 0;
pub const RADIX_SORT: i32 = 1;
pub const COUNTING_SORT: i32 = 2;
pub const INTRO_SORT: i32 = 3;
pub const TODOS: i32 = 4;

pub const INPUT1: &str = "inputs/input1.dat";
pub const INPUT2: &str = "inputs/input2.dat";
pub const INPUT3: &str = "inputs/input3.dat";
pub const INPUT4: &str = "inputs/input4.dat";
pub const INPUT5: &str = "inputs/input5.dat";
pub const INPUT6: &str = "inputs/input6.dat";

#[derive(Clone)]
pub struct Problema {
    pub lista: Vec<i32>,
    pub tamanho: usize,
    pub nome_input: String,
    pub algoritmo_usado: i32,
    pub quantidade_execucoes: i32,
}

pub fn ler_input(prob: &mut Problema, filename: &str) {
    let file = File::open(filename).unwrap_or_else(|_| {
        println!("\x1b[31m Erro ao abrir o arquivo.");
        println!(" Arquivo não encontrado ou inacessível: {} \x1b[0m", filename);
        process::exit(1);
    });

    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    // Lê o tamanho
    let tamanho: usize = lines
        .next()
        .unwrap()
        .unwrap()
        .trim()
        .parse()
        .unwrap();

    prob.tamanho = tamanho;
    prob.lista = Vec::with_capacity(tamanho);

    // Lê TODOS os números (independente de linha)
    for line in lines {
        if let Ok(linha) = line {
            let numeros = linha
                .split_whitespace()
                .filter_map(|s| s.parse::<i32>().ok());

            prob.lista.extend(numeros);
        }
    }
}

pub fn obter_nome_input(input: i32) -> String {
    if input == ESCOLHER_EM_EXECUCAO {
        loop {
            println!(" Escolha o Input a ser usado:");
            println!("1. Input 1\n2. Input 2\n3. Input 3\n4. Input 4\n5. Input 5\n6. Input 6");
            print!("Resposta: ");
            std::io::Write::flush(&mut std::io::stdout()).unwrap();

            let mut input_str = String::new();
            std::io::stdin().read_line(&mut input_str).unwrap();
            let choice: i32 = input_str.trim().parse().unwrap_or(0);

            if choice >= 1 && choice <= 6 {
                return get_nome_input(choice);
            } else {
                println!("\n \x1b[31m Opção inválida. Por favor, escolha um número entre 1 e 6. \x1b[0m \n");
            }
        }
    } else {
        get_nome_input(input)
    }
}

fn get_nome_input(input: i32) -> String {
    match input {
        1 => INPUT1.to_string(),
        2 => INPUT2.to_string(),
        3 => INPUT3.to_string(),
        4 => INPUT4.to_string(),
        5 => INPUT5.to_string(),
        6 => INPUT6.to_string(),
        _ => INPUT1.to_string(),
    }
}

pub fn obter_algoritmo_usado(algoritmo: i32) -> i32 {
    if algoritmo == ESCOLHER_EM_EXECUCAO {
        loop {
            println!("\n Escolha o algoritmo:");
            println!("1. Radix\n2. Counting\n3. intro Sort\n4. Todos");
            print!("Resposta: ");
            std::io::Write::flush(&mut std::io::stdout()).unwrap();

            let mut input_str = String::new();
            std::io::stdin().read_line(&mut input_str).unwrap();
            let choice: i32 = input_str.trim().parse().unwrap_or(0);

            if choice >= 1 && choice <= 4 {
                return choice;
            } else {
                println!("\n \x1b[31m Opção inválida.\x1b[0m\n");
            }
        }
    } else {
        algoritmo
    }
}

pub fn obter_quantidade_execucoes(quantidade_execucoes: i32) -> i32 {
    if quantidade_execucoes == ESCOLHER_EM_EXECUCAO {
        loop {
            print!("\n Digite a quantidade de execuções (1 ou mais): ");
            std::io::Write::flush(&mut std::io::stdout()).unwrap();

            let mut input_str = String::new();
            std::io::stdin().read_line(&mut input_str).unwrap();
            let choice: i32 = input_str.trim().parse().unwrap_or(0);

            if choice >= 1 {
                return choice;
            } else {
                println!("\n \x1b[31m Quantidade inválida.\x1b[0m\n");
            }
        }
    } else {
        quantidade_execucoes
    }
}

pub fn criar_problema(input: i32, algoritmo: i32, quantidade_execucoes: i32) -> Problema {
    let nome_input = obter_nome_input(input);
    let algoritmo_usado = obter_algoritmo_usado(algoritmo);
    let quantidade_execucoes_obtida = obter_quantidade_execucoes(quantidade_execucoes);

    let mut prob = Problema {
        lista: Vec::new(),
        tamanho: 0,
        nome_input: nome_input.clone(),
        algoritmo_usado,
        quantidade_execucoes: quantidade_execucoes_obtida,
    };

    ler_input(&mut prob, &nome_input);

    prob
}

pub fn criar_output(prob: &Problema, lista_tempos: &[f64]) {
    let nome_algoritmo = match prob.algoritmo_usado {
        RADIX_SORT => "Radix Sort",
        COUNTING_SORT => "Counting Sort",
        INTRO_SORT => "introsort Sort",
        _ => "Desconhecido",
    };

    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();

    let nome_arquivo = format!("outputs/output{}.dat", timestamp);

    let mut file = File::create(&nome_arquivo).unwrap_or_else(|_| {
        println!("Erro ao criar o arquivo de saída.");
        process::exit(1);
    });

    writeln!(file, "Linguagem: RUST").unwrap();
    writeln!(file, "Algoritmo: {}", nome_algoritmo).unwrap();
    writeln!(file, "Input: {}", prob.nome_input).unwrap();
    writeln!(file, "Tempos de execução (em segundos):").unwrap();

    for (i, tempo) in lista_tempos.iter().enumerate() {
        writeln!(file, "Execução {}: {}", i, tempo).unwrap();
    }
}

/*
pub fn exibir_lista(lista: &[i32]) {
    for (i, &item) in lista.iter().enumerate() {
        if i < lista.len() - 1 {
            print!("{}, ", item);
        } else {
            println!("{}", item);
        }
    }
}
*/